### PR TITLE
Add support for external tables with qualified names

### LINKED ISSUE #12607: `CREATE EXTERNAL TABLE` does not support schema (fully qualified names)
### Describe the bug

DataFusion will create external tables with qualified names as though they were single names

For example, this statement will create a table name `"staging.foo"` in the default schema, rather than a table named `foo` in the `staging` schema

```sql
create external table staging.foo stored as csv location 'foo.csv';
```

### To Reproduce

Create foo.csv:
```shell
echo "x" > foo.csv
echo "1" >> foo.csv
```

Run SQL:

```sql
DataFusion CLI v42.0.0
> create schema staging;
0 row(s) fetched.
Elapsed 0.001 seconds.

> create external table staging.foo stored as csv location 'foo.csv';
0 row(s) fetched.
Elapsed 0.011 seconds.

> select * from staging.foo;
Error during planning: table 'datafusion.staging.foo' not found
> select * from "staging.foo";
+---+
| x |
+---+
| 1 |
+---+
1 row(s) fetched.
Elapsed 0.005 seconds.
```

Note the table is in the default (`public`) schema:

```sql
> show tables;
+---------------+--------------------+-------------+------------+
| table_catalog | table_schema       | table_name  | table_type |
+---------------+--------------------+-------------+------------+
| datafusion    | public             | staging.foo | BASE TABLE |
| datafusion    | information_schema | tables      | VIEW       |
| datafusion    | information_schema | views       | VIEW       |
| datafusion    | information_schema | columns     | VIEW       |
| datafusion    | information_schema | df_settings | VIEW       |
| datafusion    | information_schema | schemata    | VIEW       |
+---------------+--------------------+-------------+------------+
6 row(s) fetched.
Elapsed 0.002 seconds.
```

### Expected behavior

The external table should be named `foo` in the `staging` schema

Note this works as expected for `CREATE TABLE` (just not `CREATE EXTERNAL TABLE`)

```sql
> create table staging.bar as values (2);
0 row(s) fetched.
Elapsed 0.006 seconds.

> select * from staging.bar;
+---------+
| column1 |
+---------+
| 2       |
+---------+
1 row(s) fetched.
Elapsed 0.001 seconds.

> show tables;
+---------------+--------------------+-------------+------------+
| table_catalog | table_schema       | table_name  | table_type |
+---------------+--------------------+-------------+------------+
| datafusion    | public             | staging.foo | BASE TABLE |
| datafusion    | staging            | bar         | BASE TABLE |
| datafusion    | information_schema | tables      | VIEW       |
| datafusion    | information_schema | views       | VIEW       |
| datafusion    | information_schema | columns     | VIEW       |
| datafusion    | information_schema | df_settings | VIEW       |
| datafusion    | information_schema | schemata    | VIEW       |
+---------------+--------------------+-------------+------------+
7 row(s) fetched.
Elapsed 0.004 seconds.
```

### Additional context

Found by @matthewmturner  in https://github.com/datafusion-contrib/datafusion-dft/issues/154#issuecomment-2371404530

### PR DESCRIPTION
**Pull request by OussamaSaoudi**
## Which issue does this PR close?

Closes #12607

## What changes are included in this PR?

This PR adds the ability to create external tables with qualified names.
## Are these changes tested?

I add a unit  test to verify the expected behaviour in the issue. Namely, a call to `create external table staging.foo` should be a `TableReference::Partial {schema: "staging", table: "foo"}` qualified reference.

I also add two sqllogic tests to check happy path and error case where there is no schma that the qualified name maps to.
## Are there any user-facing changes?

Yes there are user-facing changes. Previously, external tables with qualified names was unsupported.

### CODE REVIEW

File: datafusion/core/src/catalog_common/mod.rs [hunk 1]
```diff
@@ -185,9 +185,8 @@ pub fn resolve_table_references(
                 let _ = s.as_ref().visit(visitor);
             }
             DFStatement::CreateExternalTable(table) => {
-                visitor
-                    .relations
-                    .insert(ObjectName(vec![Ident::from(table.name.as_str())]));
+                let idents: Vec<Ident> = table.name.split('.').map(Ident::from).collect();
```
Comment (OussamaSaoudi) [inline datafusion/core/src/catalog_common/mod.rs:188]: This part is not necessary for the bug fix. Not sure if I should leave it here to maintain consistency of external table name handling.

**Comment by OussamaSaoudi** [comment]
@alamb I've never written tests for Datafusion before. What's a good spot to test this change?

File: datafusion/core/src/catalog_common/mod.rs (hunk 1 above)
Comment (jonahgao) [reply datafusion/core/src/catalog_common/mod.rs:188]: I think the type of [table.name](https://github.com/apache/datafusion/blob/3892499b28135f9862ffac3e4f19524bb5b77c59/datafusion/sql/src/parser.rs#L184) needs to be changed from `String` to `ObjectName`.

File: datafusion/sql/src/statement.rs [hunk 2]
```diff
@@ -1239,8 +1240,9 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
         let ordered_exprs =
             self.build_order_by(order_exprs, &df_schema, &mut planner_context)?;
 
-        // External tables do not support schemas at the moment, so the name is just a table name
-        let name = TableReference::bare(name);
+        let idents: Vec<Ident> = name.split('.').map(Ident::from).collect();
+        let name = object_name_to_table_reference(ObjectName(idents), false)
+            .unwrap_or(TableReference::bare(name));
```
Comment (jonahgao) [inline datafusion/sql/src/statement.rs:1245]:
```suggestion
        let name = self.object_name_to_table_reference(name)?;
```
We should respect the `enable_ident_normalization` option here

**Comment by alamb** [comment]
> @alamb I've never written tests for Datafusion before. What's a good spot to test this change?

The best tests I think would be in sqllogictests

Here are the instructions: https://github.com/apache/datafusion/tree/main/datafusion/sqllogictest

Ideally you should be able to extend one of the existing test files in https://github.com/apache/datafusion/tree/main/datafusion/sqllogictest/test_files 

Perhaps this one:
https://github.com/apache/datafusion/blob/main/datafusion/sqllogictest/test_files/create_external_table.slt

**Comment by alamb** [comment]
It seems like the Ci is failing here. Marking as draft as that is addressed. Thank you

**Comment by alamb** [comment]
Is this one ready for review @OussamaSaoudi ?

File: datafusion/sql/src/statement.rs [hunk 3]
```diff
@@ -1239,8 +1239,7 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
         let ordered_exprs =
             self.build_order_by(order_exprs, &df_schema, &mut planner_context)?;
 
-        // External tables do not support schemas at the moment, so the name is just a table name
```
Comment (alamb) [inline datafusion/sql/src/statement.rs:1242]: 🎉

**Review by alamb — APPROVED** [review]
Looks great to me @OussamaSaoudi  -- thank you for the contribution

**Comment by alamb** [comment]
FYI @matthewgapp

**Comment by OussamaSaoudi** [comment]
Sorry for the delay on marking it ready to review, I'd planned on double checking it before review. Looks like it went smooth though lol :D

**Comment by matthewmturner** [comment]
@OussamaSaoudi thanks for this!

**Comment by alamb** [comment]
> Sorry for the delay on marking it ready to review, I'd planned on double checking it before review. Looks like it went smooth though lol :D

Indeed -- it looks great. Thank you again!

**Pull request merged**